internal/cache/lru.go

```diff
@@ -0,0 +1,119 @@
+package cache
+
+import (
+	"sync"
+)
+
+type cacheNode struct {
+	Key   chacheKey
+	Value []byte
+	Next  *cacheNode
+	Prev  *cacheNode
+}
+
+// LRUCache is a cache with the least-recently-used policy.
+type LRUCache struct {
+	hashMap  map[chacheKey]*cacheNode
+	start    *cacheNode
+	end      *cacheNode
+	count    int
+	capacity int
+	lock     sync.Mutex
+}
+
+// NewLRUCache creates a new LRUCache instance with the given capacity.
+func NewLRUCache(capacity int) *LRUCache {
+	return &LRUCache{
+		hashMap:  make(map[chacheKey]*cacheNode, capacity),
+		capacity: capacity,
+	}
+}
+
+// Put adds a new key-value pair to the cache. If the cache is full then
+// the least-recently-used key-value pair is removed.
+func (c *LRUCache) Put(k []byte, v []byte) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// Build the cache key and node
+	key := newCacheKey(k)
+	node := &cacheNode{Key: key, Value: v}
+
+	// Cache is empty
```

Comment:
Probably unnecessary. Otherwise, please expand with what happens when cache is empty